### PR TITLE
Support GIN operator classes and JSON path expressions

### PR DESCRIPTION
**Pull request by OrangeDog**
Example fixes for https://code.djangoproject.com/ticket/28077 and https://code.djangoproject.com/ticket/28928

### CODE REVIEW

**Comment by charettes** [comment]
I think we'll want to keep using the `field__tags__0` syntax here and not a specialized dot separated one as that's how `JSONField`/`HStoreField` are currently queryable anyway.

**Comment by charettes** [comment]
Also this has a bit of overlap with adding support for expression in fields which is tracked in https://code.djangoproject.com/ticket/26167 and tentatively implemented in #8056.

`GINIndex` shouldn't assume the fields it's bound to is a `JSONField` and generate such lookups; `KeyTransform`/`KeyTextTransform` should be provided in fields instead and they should be in charge of generating the adequate SQL.

File: django/contrib/postgres/indexes.py [hunk 1]
```diff
@@ -53,17 +53,40 @@ def get_with_params(self):
...
+        columns.quote_name = lambda x: x
+        for field, path in zip(self.fields, self.paths):
+            if not path:
+                expr = quote_name(field)
+            else:
+                expr = '({} -> {})'.format(
+                    quote_name(field),
+                    ' -> '.join("'{}'".format(key) for key in path))
+            columns.rename_column_references(
+                model._meta.db_table, field, expr)
```
Comment (charettes) [inline django/contrib/postgres/indexes.py]: This should be handled by the fields themselves, what you've written here is what a special cased `JSONFieldGinIndex.create_sql` implementation would look like.

**Comment by OrangeDog** [comment]
@charettes that's fine. I wasn't expecting this to actually be merged.
I just wanted to provide some code showing how I was dealing with it, hopefully to help people doing it properly.

**Pull request closed without merging**